\documentclass[a4]{bioinfo}
\usepackage{url}
\usepackage{acrmbio}
\usepackage{epsfig}
\usepackage{tabularx}

\bibliographystyle{bioinformatics}
%\bibliographystyle{natbib}
\emergencystretch 2in

\copyrightyear{2014}
\pubyear{2014}
\title[JSAV: JavaScript Sequence Alignment Viewer]{Application Note:\\
JSAV: JavaScript Sequence Alignment Viewer}
\date{26th June, 2014}
\author[Martin]{Andrew C. R. Martin\mbox{${}^{\rm a}$}\footnote{to whom
    correspondence should be addressed}}
\address{\mbox{${}^{\rm a}$}Institute of Structural and Molecular Biology, 
Division of Biosciences,
University College London, Darwin Building, Gower Street, London WC1E 6BT}
\history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}

\editor{Associate Editor: XXXXXXX}

\begin{document}
\maketitle

\begin{abstract}
\noindent{\bfseries Summary:} The JavaScript Sequence Alignment Viewer
(JSAV) is designed as a simple-to-use JavaScript component for
displaying sequence alignments on web pages.  The display of sequences
is highly configurable with options to allow alternative colouring
schemes, sorting of sequences and `dotifying' repeated amino acids. An
option is also available to submit selected sequences to another web
site, or to other JavaScript code.  \\
\noindent{\bfseries Availability and Implementation:} 
JSAV is implemented purely in JavaScript making use of JQuery and
JQuery-UI. It does not use any HTML5 specific options to help with
browser compatibility. The code is documented using JSDOC
and is available from \url{http://www.bioinf.org.uk/software/jsav/}\\
\noindent{\bfseries Contact:} andrew@bioinf.org.uk --or--
andrew.martin@ucl.ac.uk
\end{abstract}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Introduction}
Viewing multiple sequence alignments (MSAs) is a fundamental
requirement in analysis of protein sequences, allowing us to visualize
conservation across protein families as well as unusual features of
particular sequences. As a result, there are a plethora of tools for
viewing MSAs. These range from tools which provide attractive printed
output, through (operating-system dependent, or independent)
standalone tools, to web-based viewers.

Two of the earliest tools were HOMED\cite{stockwell:homed} and
MALIGNED\cite{clark:maligned} written for VAX/VMS workstations.
Neither seems to be actively maintained or easily available any
more. Other early viewers include GeneDoc\cite{nicholas:genedoc},
BioEdit, Seaview\cite{galtier:seaview} and DCSE\cite{derijk:dcse}
which is part of the RnaViz package for visualizing RNA secondary
structure\cite{derijk:rnaviz}, but which can be used for protein
sequence alignments.  A problem in writing graphical software is the
operating-system dependency of many graphics
libraries. CINEMA\cite{parrysmith:cinema} was probably the first
sequence alignment viewer and editor implemented in Java, a platform
independent programming language allowing graphical user interfaces
(GUIs) to run on any operating system. It has now been rewritten in
C++ and is part of UTOPIA\cite{pettifer:utopia}. Other software
includes MPSA\cite{blanchet:mpsa}, ANTHEPROT\cite{deleage:antheprot}
and ClustalX\cite{thompson:clustalx}, a GUI for the ClustalW multiple
sequence alignment program, providing an integrated environment for
aligning sequences and analyzing results. Clustal Omega is the most
recent version, but at the time of writing only has a command line
interface --- a beta version of a GUI is due to be released soon.

More recent developments include the Protein Family Alignment
Annotation Tool (PFAAT)\cite{johnson:pfaat} designed specifically for
family analysis and incorporating residue annotation tools as well as
integration with Jmol for protein structure display. Like early
versions of CINEMA, PFAAT is implemented in Java for operating system
independence. CLC~Viewer is a recent free package written in Java
which contains a number of integrated tools and acts as a core product
for adding other features through a commercial version.  A more
complete list of MSA viewers is available on the web at
\url{http://en.wikipedia.org/wiki/List_of_alignment_visuzlization_software}.

Probably the most popular of the available tools is
Jalview\cite{clamp:jalview} which is available in two versions: a
standalone Java application which provides many tools and facilities,
and as a `light' version (JalviewLight) --- a Java applet that can be
embedded in a web page. The latter responds to the need for web site
developers to be able to embed MSA visualization into web pages.

However, in recent years there has been a gradual move away from using
Java applets in web development. New HTML features such as the HTML5
Canvas, and powerful JavaScript libraries such as Bootstrap, JQuery
and JQuery-UI that provide an easier syntax for accessing elements of
a web page together with new widgets such as sliders and drag-and-drop
support, have overtaken Java as the method of choice for creating
interactive web sites with complex requirements.  Such features are
used widely by popular web sites such as Google Mail, Google Docs and
Facebook.  Recently the Jmol structure viewer has been reimplemented
in JavaScript as JSmol.

To our knowledge, while there is an intention to port JalviewLight to
JavaScript, there are currently no JavaScript based sequence alignment
viewers available. To address this, and our own requirements to embed
MSA visualization into web pages, we have developed JSAV (JavaScript
Sequence Alignment Viewer). JSAV is designed to be a lightweight
component that can easily be dropped into a web site to display and
manipulate a sequence alignment. It allows colouring, sorting and
removing sequences as well as integration with other web pages and
local JavaScript code.



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Implementation}

\begin{figure}
\footnotesize
\begin{verbatim}
var MySeqs = Array();
MySeqs.push({ id :"id1b1.L",  
   sequence :"SASSSVNYMYACREFGHIKLMNPTRSTVWY"});
MySeqs.push({ id :"id1a.L",   
   sequence :"SASSSTNYMYACDEFGHIKLMNPQRSTVWY"});
MySeqs.push({ id :"id2b1.L",  
   sequence :"SASSTCNYMTACDEEGHIKLMNP-RSTCWY"});

var MyOptions = Array();
MyOptions.sortable = true;
MyOptions.selectable = true;
MyOptions.deletable = true;
MyOptions.toggleDotify = true;
MyOptions.toggleNocolour = true;
MyOptions.consensus = true;
MyOptions.selectColour = true;

printJSAV('sequenceDisplay', MySeqs, MyOptions);
\end{verbatim}
\caption{\label{fig:code}Example code to create a
JavaScript array of sequence objects, set options and 
create the alignment viewer.}
\end{figure}

\begin{figure}
\epsfig{file=demo.eps,width=\columnwidth}
\caption{\label{fig:demo}A typical JSAV alignment view}
\end{figure}

%JSAV is implemented purely in JavaScript. JSAV employs the JQuery
%library to ease access to elements of the HTML that it generates and
%uses JQuery-UI to implement a two-value slider that is used to specify
%a range of positions in the alignment.

JSAV is implemented purely in JavaScript employing the JQuery and
JQuery-UI libraries. As input, the code requires an array of
JavaScript objects which contain two elements: a unique identifier for
a sequence and the sequence itself --- all sequences must be
pre-aligned.  Secondly a set of options can be provided. These control
the display and facilities available to the end user of a web site to
modify the view of the MSA.  A brief extract of sample code is shown
in Fig~\ref{fig:code} with the results shown in Fig~\ref{fig:demo}.

The end-user can sort the sequences --- the code selects the most
representative sequence, displaying that at the top of the alignment
followed by the most similar sequence and so on. By default, sorting
is performed across the whole sequence, but a two-handled slider
allows the range of positions on which the sort is based to be
modified. Different colouring schemes are available duplicating those
provided in Jalview. The alignment can also be `dotified', replacing
residues repeated between sequences with dots in order to emphasize
amino acid differences. Colouring of dotified residues can also be
switched off or on. Sequences can be selected and deleted from the
alignment; a consensus sequence can be displayed at the bottom of the
alignment and updates automatically when sequences are deleted. The
complete set of sequences, or a selected subset, can be submitted to
another web site, or passed to another JavaScript function for
integration with other tools. Tooltips are provided for each option
and all options are documented in detail on the web site. The number
and size of sequences in the MSA is limited only by the memory
available to the web browser.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Availability}
The software may be downloaded from
\url{http://www.bioinf.org.uk/software/jsav/} where demonstrations,
including the ability to upload your own MSA, are available together
with full documentation implemented with JSDOC.

\bibliography{jsav}

\end{document}
